### PR TITLE
Fix reconnection issue

### PR DESCRIPTION
**Pull request by ms42**
Changes:
* Fix timer thread, so that it won't die on an exception and keeps firing ClientReconnecting or ClientDisconnected events (in rare occasions `connectionState()` throws a `NullPointerException`)
* Add safe guard when using `isRunning()` for either creating or destroying variable `timer`
* Add settings parameter `disconnectOnTcpChannelError` to switch off immediate disconnection in case of a channel error, like a connection reset by the server (which happens quite often on a reconnect using es v4).

This should fix issue #29. Testing it with the following configuration and having no issues so far:
```
EventStoreBuilder.newBuilder()
                    .singleNodeAddress("127.0.0.1", 1113)
                    .userCredentials("admin", "changeit")
                    .maxOperationRetries(-1)
                    .maxReconnections(-1)
                    .disconnectOnTcpChannelError(false)
                    .build();
```

### CODE REVIEW

**Comment by msemys** [comment]
Thanks for PR!

I really liked connect/disconnect and timer fixes. It's a good idea to have `disconnectOnTcpChannelError` switch at the moment, because I was thinking about to remove that `.whenChannelError(EventStoreTcp.this::onChannelError)` but I need more testing on this.

Could you resolve the conflict in `EventStoreTcp.java`?

about `ProjectionException: Server returned 408 (Server was unable to handle request in time)` - it's likely related to load and environment where es is running.
I was playing with es v5 a couple of days and my first impression is that integration tests with es v5 runs much stable, than with es v4.

File: src/main/java/com/github/msemys/esjc/EventStoreTcp.java [hunk 1]
```diff
@@ -688,32 +692,40 @@ private void onBadRequest(TcpPackage tcpPackage) {
     }
 
     private void onChannelError(Throwable throwable) {
-        handle(new CloseConnection("Error when processing TCP package", throwable));
+        if(settings().disconnectOnTcpChannelError) {
+            handle(new CloseConnection("Error when processing TCP package", throwable));
+        } else {
+            logger.debug("Failed processing TCP package", throwable);
```
Comment (msemys) [inline src/main/java/com/github/msemys/esjc/EventStoreTcp.java]:
```suggestion
            logger.error("Failed processing TCP package", throwable);
```

File: src/main/java/com/github/msemys/esjc/EventStoreTcp.java [hunk 2]
```diff
@@ -688,32 +692,40 @@ private void onBadRequest(TcpPackage tcpPackage) {
...
-            case CONNECTED:
-                checkOperationTimeout();
-                break;
+                    break;
+                case CONNECTED:
+                    checkOperationTimeout();
+                    break;
+            }
+        } catch (Exception e) {
+            logger.debug("Caught exception to keep timer thread running", e);
```
Comment (msemys) [inline src/main/java/com/github/msemys/esjc/EventStoreTcp.java]:
```suggestion
            logger.error("Error occurred in timer thread", e);
```

File: src/main/java/com/github/msemys/esjc/Settings.java [hunk 3]
```diff
@@ -569,6 +589,10 @@ public Settings build() {
                 failOnNoServerResponse = false;
             }
 
+            if (disconnectOnTcpChannelError == null) {
+                disconnectOnTcpChannelError = true;
```
Comment (msemys) [inline src/main/java/com/github/msemys/esjc/Settings.java]: Lets use `false` as default. Currently I don't see the benefit of closing connection when client encounters an error...

```suggestion
                disconnectOnTcpChannelError = false;
```

Comment (ms42) [reply src/main/java/com/github/msemys/esjc/Settings.java]: I'm very happy to use `false` as default - I just used the existing behaviour as default, but I agree that changing it to `false` would be more useful. I'll update the Javadoc as, well.

**Comment by ms42** [comment]
With changing the default behaviour from `true` to `false` for `disconnectOnTcpChannelError` two integration tests fail:
1. `ITSslCertificateConnection..failsWithNonMatchingCertificateFile`
2. `ITSslCommonNameConnection.failsWithNonMatchingCommonName`

First guess without taking a deeper look into the failing tests is that the tests are running into a timeout as the connection is not disconnected immediately and the client keeps trying to reconnect.

File: src/main/java/com/github/msemys/esjc/EventStoreBuilder.java [hunk 4]
```diff
@@ -452,6 +453,18 @@ public EventStoreBuilder failOnNoServerResponse(boolean failOnNoServerResponse)
...
     }
 
+    /**
+     * Sets whether or not to disconnect the client on detecting a channel error. By default, it is enabled and
+     * client disconnects immediately. If it is disabled the client tries to reconnect according to {@link #maxReconnections(int)}.
+     *
+     * @param disconnectOnTcpChannelError {@code true} to disconnect or {@code false} to try to reconnect.
+     * @return the builder reference
+     */
+    public EventStoreBuilder disconnectOnTcpChannelError(boolean disconnectOnTcpChannelError) {
```
Comment (msemys) [inline src/main/java/com/github/msemys/esjc/EventStoreBuilder.java]: could you please update the javadocs

Comment (ms42) [reply src/main/java/com/github/msemys/esjc/EventStoreBuilder.java]: Good catch, sorry I missed this one :see_no_evil:

**Comment by msemys** [comment]
:+1: thanks!

**Pull request merged**